STAN Models for bbsBayes

steffilazerte opened the pull request:
This pull request contains extensive development to include (among other things) models built with STAN rather than JAGS. 

Current framework to ensure that functionality is maintained and we (hopefully) do not introduce bugs!

- Tidying existing functions to ensure consistency (`XXX_tidy()`)
- Modifying tidied functions to use Stan model outputs (get the original function name `XXX()`)
- Once finished, original and tidied functions will be omitted (but right now are `XXX_orig()`)

> **Stage:** Cleaning and Scrubbing
> Working to clean up existing code, solidify checks and tests and add documentation

### Progress

- [x] New core functionality
  - [x] Update functions to handle Stan models via `cmdstanr`
    - [X] `generate_trends()`, `generate_indices()`, `plot_map()` (was `generate_map()`), `plot_geofacet()` (was `geofacet_plot()`), `plot_indices()`
    - [x] `model_to_file()`
  - [X] Add functionality for spatial modelling (i.e. `prepare_spatial()`)
  - [X] Add convergence checks (i.e., `convergence()`)
  - [X] Add option to provide custom stratification
- [x] Best Practices
  - [x] Function tests and code coverage indicators
  - [X] Tidying
  - [X] Deprecate/defunct functions and arguments gracefully
  - [x] Best practices ([`goodpractices`](https://cran.r-project.org/web/packages/goodpractice/index.html))
  - [x] Passing checks
- [x] Last to-dos
  - [x] n_route and n_routes_total should update if strata excluded
  - [x] obs_mean should be zero in indices if obs were missing in prior years
- [x] Documentation
  - [x] Update function references
  - [x] Vignettes
    - [x] Stratification / Data prep
    - [x] Model Choice
    - [x] Results and summaries (annual indices, trend-estimates and maps)
- [ ] Ready for Release
  - [x] Update sample data to 2020 (update tests accordingly)
  - [x] Checks (notes etc. )
  - [x] Deal with location of BBS data (use CRAN best practices)
  - [x] Deal with location of compiled model files
  - [x] Deal with any potentially fragile or long running tests
  - [x] Deal with any potentially fragile or long running examples
  - [ ] Host on R-Universe

steffilazerte committed:
Initial set of changes
- Creating tidy versions of existing functions
- Creating stan versions of functions
- Creating test scripts
- Add STAN Models
- Basic metadata updates
- Basic setup vignette stub

steffilazerte committed:
Clean up and fix indexing issue with tibbles

steffilazerte committed:
Cleanup

steffilazerte commented:
@AdamCSmithCWS I'm working on getting the STAN models all working with their options (then I'll tackle the spatial models). The jags models seem to include four (slope, firstdiff, gam, gamye), but the original stan models only two (non-spatial: slope and gamye).

Will there be four regular STAN models eventually? Or are the firstdiff and gam models either not relevant or incorporated into the existing models?

steffilazerte committed:
Finalize data prep and tidy

steffilazerte committed:
Arrange data to simplify comparisons

steffilazerte committed:
Create run_model for stan

steffilazerte committed:
Add checks

steffilazerte committed:
Tidy data stratification

steffilazerte committed:
Finalize generation of indices

steffilazerte committed:
Add latin1 encoding to fix existing fetch_bbs_data() function

steffilazerte committed:
Add tidy fetch_bbs_data()

steffilazerte committed:
Add more checks to stratify_tidy

AdamCSmithCWS commented:
> @AdamCSmithCWS I'm working on getting the STAN models all working with their options (then I'll tackle the spatial models). The jags models seem to include four (slope, firstdiff, gam, gamye), but the original stan models only two (non-spatial: slope and gamye).
> 
> Will there be four regular STAN models eventually? Or are the firstdiff and gam models either not relevant or incorporated into the existing models?

Yes. Sorry. I will push versions of the firstdiff and gam two models within the next two days.

AdamCSmithCWS commented:
> > @AdamCSmithCWS I'm working on getting the STAN models all working with their options (then I'll tackle the spatial models). The jags models seem to include four (slope, firstdiff, gam, gamye), but the original stan models only two (non-spatial: slope and gamye).
> > Will there be four regular STAN models eventually? Or are the firstdiff and gam models either not relevant or incorporated into the existing models?
> 
> Yes. Sorry. I will push versions of the firstdiff and gam two models within the next two days.

@steffilazerte: 
My [bbsStanBayes repo](https://github.com/AdamCSmithCWS/bbsStanBayes) now includes the first difference models (spatial and non). Note: they won't generate estimates that directly match the earlier version of the bbsBayes firstdifference models. These versions are hierarchical versions that share information among strata, the earlier versions were not. This represents an elaboration (improvement) on the earlier models and I'm not sure it's worth adding the original version. Perhaps we should discuss this. If we were to add the original non-hierarchical version, there would have to be 3 first-difference models...

I will add the gam models tomorrow.

AdamCSmithCWS commented:
> 

All of the bbsBayes models are now translated to Stan in the bbsStanBayes repo.
In addition, I added a version of the first difference model that matches the USGS implementation. I've called it a [NonHierarchical version ](https://github.com/AdamCSmithCWS/bbsStanBayes/blob/main/models/first_difference_NonHier_bbs_CV.stan)of the model. The code to run it is now in the same [R script as for the other first difference models](https://github.com/AdamCSmithCWS/bbsStanBayes/blob/main/Fit_first_difference_models.R). It includes a boolean object ` Non_hierarchical <- TRUE ` object defined in the script that sets everything up correctly.

steffilazerte committed:
Fetch/Load data
- Stop on exists (not warn) and add meta data
- tidy version of loading
- using rds

steffilazerte committed:
Fix final stratify discrepancies - correct french species, use inner join, be explicit about rid

steffilazerte committed:
Clean up

steffilazerte committed:
Add species check

steffilazerte committed:
Add checks, fix and tests for prepare_data()

steffilazerte committed:
Start vignette

steffilazerte committed:
Fix incorrect order in indices prov_state

steffilazerte committed:
Update docs
- Defunct deprecated functions
- Add package-level documentation

steffilazerte committed:
Tidy get species aou

steffilazerte committed:
Rename to _orig / _tidy and stan or final functions have no suffix

steffilazerte committed:
Function to create init defs

steffilazerte committed:
Clarify models and init_def
- file names and variants
- naming with snakecase
- all non-spatial variables

steffilazerte committed:
Move model parameters from prepare_data to run_model and finalize spatial models

steffilazerte committed:
Modify stan models to match naming conventions for spatial

steffilazerte committed:
Use different encoding on differnet systems

steffilazerte committed:
Checks for stratify

steffilazerte committed:
Add deps

steffilazerte committed:
Don't commit compiled Stan models

steffilazerte committed:
Modify storage of bbs data
- Allow storing of all types (level adn release)
- Function to remove data

steffilazerte committed:
Standardize names

steffilazerte committed:
Clarify spatial checks and match argument names to cmdstanr

steffilazerte committed:
Correct checks

steffilazerte committed:
remove extract_index_data()

steffilazerte committed:
Basic tidying and updating docs, add convergence placeholder

steffilazerte committed:
Add pkgdown and test articles

steffilazerte committed:
Update tests

steffilazerte committed:
Add preliminary RELEASE and other housekeeping

steffilazerte committed:
Add rpid back in to join to mirror original

steffilazerte committed:
Rename arguments and function names

steffilazerte committed:
Revert to original sample data

steffilazerte committed:
update names

steffilazerte committed:
Use consistent meta_data

steffilazerte committed:
Update tests and docs

steffilazerte committed:
Standardizing attributes and arguments

steffilazerte committed:
Start common docs

steffilazerte committed:
Use interal loading function

steffilazerte committed:
Clarify metadata

steffilazerte committed:
Cleanup and general tidying

steffilazerte committed:
Update docs

steffilazerte committed:
Missing }

steffilazerte committed:
Add checks and save output

steffilazerte committed:
add more checks

steffilazerte committed:
Add custom stratification and move species to stratify()

steffilazerte committed:
Combine loading sample and full data

steffilazerte committed:
Simplify load_maps() and add options for base maps

steffilazerte committed:
Combined species forms in fetch_bbs_data

steffilazerte committed:
Add option of custom strata to indices and trends and corresponding plots

steffilazerte committed:
Add more checks

steffilazerte committed:
Add helper functions

steffilazerte committed:
housekeeping

steffilazerte committed:
Update docs

steffilazerte committed:
Update tests

steffilazerte committed:
Add draft vignettes

steffilazerte committed:
Add convergence()

steffilazerte committed:
Update design notes

steffilazerte committed:
housekeeping

steffilazerte committed:
Fix changes to spatial

steffilazerte committed:
Fix inconsistencies

steffilazerte committed:
Omit unused functions

steffilazerte committed:
Omit old functions from docs and example tests

steffilazerte committed:
Fix inconsistencies

steffilazerte committed:
add dev option for use in creating strata

steffilazerte committed:
Add model example (?)

steffilazerte committed:
housekeeping

steffilazerte committed:
Add more params to meta, allow all cmdstanr args and don't use example

steffilazerte committed:
Updat tests

steffilazerte committed:
Fix vignettes and create pkgdown on GitHub Actions

steffilazerte committed:
Ignore intermediate files in vignettes

steffilazerte committed:
Check for bbs data

steffilazerte committed:
omit old indices docs

steffilazerte committed:
try installing bbs bayes before runnin pkgdown

steffilazerte committed:
Add internal function without interactive terms for use in testing and GHA

steffilazerte committed:
Clean up dependencies and have GHA install cmdstan

steffilazerte committed:
Fix first round of check, missing deps, etc.

steffilazerte committed:
Move model exec to cache and remove jags models

steffilazerte committed:
Fix docs

steffilazerte committed:
Fix naming

steffilazerte committed:
Move base maps in data-raw and don't offer option to load, update internal maps

steffilazerte committed:
Fix data docs

steffilazerte committed:
Housekeeping

steffilazerte committed:
Add cmdstanr repo (a la oharar/LatentINLA)

steffilazerte committed:
Even better

steffilazerte committed:
Use explicit references to species_forms (otherwise can't be found)

steffilazerte committed:
add out_dir (argh)

steffilazerte committed:
Move bbs data checks to dedicated function

steffilazerte committed:
Install rnaturalearthhires for access to ne_states()

steffilazerte committed:
Don't forget rnaturalearthhires' repository

steffilazerte committed:
Try again?

steffilazerte committed:
arg

steffilazerte committed:
again

steffilazerte committed:
finally?

steffilazerte committed:
More forgotten packages

assignUser committed:
Add data caching to pkgdown workflow

steffilazerte committed:
Merge pull request #1 from assignUser/patch-1

Add data caching to pkgdown workflow

steffilazerte committed:
Dont' check for data in vignettes unless for user

steffilazerte committed:
Merge branch 'master' of github.com:steffilazerte/bbsBayes

steffilazerte committed:
Echo to output

steffilazerte committed:
Get another missing package and add missing quotes

steffilazerte committed:
Report data location in actions

steffilazerte committed:
Add message when checking bbs data, add to end of Actions to confirm location, use lazy for pkgdown testing

assignUser committed:
fix cache dir format

steffilazerte committed:
Merge pull request #2 from assignUser/patch-2

Fix cache dir format

steffilazerte committed:
Fix latlong map as well

steffilazerte committed:
Don't forget wy in slopes

steffilazerte committed:
Add n_routes

steffilazerte committed:
Update maps

steffilazerte committed:
Add basic tests for indices, trends and plots

steffilazerte committed:
Merge branch 'master' of github.com:steffilazerte/bbsBayes

steffilazerte committed:
Change generate_map and geofacet_plot to plot_map and plot_geofacet.

steffilazerte committed:
List all removed functions as defunct

steffilazerte committed:
Add startup message

steffilazerte committed:
Try again with deprecated/defunct

steffilazerte committed:
Rename test files

steffilazerte committed:
Correct neighbour filling to modify original and not to add all neighbours

steffilazerte committed:
Add label size to prepare_spatial map output

steffilazerte committed:
Omit option to have POINTs in prepare_spatial

steffilazerte committed:
Tinker with messaging for islands

steffilazerte committed:
Update prepare spatial tests

steffilazerte committed:
Add map checking to examples for spatial_data

steffilazerte committed:
Fix max_backcast and strata exclusions

steffilazerte committed:
Make convergence estimates match summary and add convenience functions

steffilazerte committed:
Housekeeping

steffilazerte committed:
Fix problems with generate_indicies and different year ranges

steffilazerte committed:
Add more checks and reorganize generate_trends

steffilazerte committed:
Organize and check incoming data

steffilazerte committed:
Update tests

steffilazerte committed:
Add codecoverage and CI checks

steffilazerte committed:
Add extra repos

steffilazerte committed:
Install cmdstan and fetch data

steffilazerte committed:
Use exact years as indices and deal with missing data

steffilazerte committed:
Install bbsBayes before checking (so can fetch data first)

steffilazerte committed:
Correct convergence function name and add other helper functions

steffilazerte committed:
Use rtools 4.2 for windows (required by cmdstanr)

steffilazerte committed:
Actually use the rtools specified

steffilazerte committed:
Update docs

steffilazerte committed:
Add check for file type

steffilazerte committed:
Remove old functions! 🥁

steffilazerte committed:
Add back get_sb_id()

steffilazerte committed:
Catch a few missing imports

steffilazerte committed:
Fix documentation and general notes

steffilazerte committed:
Fix year indices

steffilazerte committed:
Add option to use custom model file, standardize run_model variables, fix notes

steffilazerte committed:
Use example models

steffilazerte committed:
Clean up

steffilazerte committed:
Add to instructions

steffilazerte committed:
Update vignette drafts

steffilazerte committed:
Avoid note

steffilazerte committed:
Correct tests and add notes

steffilazerte committed:
Move convergence to article

steffilazerte committed:
Try strat again?

steffilazerte committed:
Ensure loading?

steffilazerte committed:
Fix CI errors?

steffilazerte committed:
Fix goodpractice checks

steffilazerte committed:
Use linewidth rather than size for ggplot2 v 3.4.0

steffilazerte committed:
Update tests for trend col names

steffilazerte committed:
Update tests for correcting backcast < first_year (rather than <= first_year)

steffilazerte committed:
Correct chuck arguments

steffilazerte committed:
Fix use of .data in tidyselect

steffilazerte committed:
Fix doc error

steffilazerte committed:
update arg name

steffilazerte committed:
Use snapshots for models, and internal model for later tests

steffilazerte committed:
Skip vignettes on CI checks (already run by pkgdown)

steffilazerte committed:
Rename test files

steffilazerte committed:
Add model util tests

steffilazerte committed:
Move load_map to utils

steffilazerte committed:
separate out helper functions from utils

steffilazerte committed:
Use new sample data

steffilazerte committed:
Clarify species

steffilazerte committed:
Ensure strata visible

steffilazerte committed:
Add utils and helper tests

steffilazerte committed:
Update tests for new sample data

steffilazerte committed:
update man

steffilazerte committed:
Add some more package check details

steffilazerte committed:
Once more with snaps!

steffilazerte committed:
Move more checks into check functions, standardize checks

steffilazerte committed:
Add more tests, skip random seed tests on CI

steffilazerte committed:
Don't build manual on CI

steffilazerte committed:
Clarify expected location

steffilazerte committed:
Add more fetch data tests

steffilazerte committed:
Rename helpers test

steffilazerte committed:
Add pkgdown structure, update workflow article

steffilazerte committed:
Skip all seed tests on CI

steffilazerte committed:
Rename model_to_file to copy_model_file

steffilazerte committed:
Add missing contents

steffilazerte committed:
Fix typo

steffilazerte committed:
Simplify REAMDE

steffilazerte committed:
Correctly document data, add to pkgdown, and omit old sample data

steffilazerte committed:
add missing models

steffilazerte committed:
Update the getting started article

steffilazerte committed:
Clarify arguments

steffilazerte committed:
Clarify which articles are articles!

steffilazerte committed:
Remove original sample data

steffilazerte committed:
Update to actual current example data

steffilazerte committed:
Mark missing years

steffilazerte committed:
Check routes change with backcast removed

steffilazerte committed:
quiet tests

steffilazerte committed:
Don't include deprecated/defunct functions

steffilazerte committed:
Add bbsBayes version to model metadata

steffilazerte committed:
Add trends info (slope and year-range) to plot

steffilazerte committed:
Update examples

steffilazerte committed:
Add test for ... arguments in run_model()

steffilazerte committed:
Add checks clarify docs

steffilazerte committed:
When removing "all" cache don't check other options

steffilazerte committed:
Quiet/fix tests

steffilazerte committed:
Update (again) with new model files

steffilazerte committed:
Add CRAN-happy data directory

steffilazerte committed:
Housekeeping

steffilazerte committed:
Add pre-compiled model vignettes

steffilazerte committed:
Add all models

steffilazerte committed:
Split part of run_model into prepare_model
- add set_seed argument for precise seeds

steffilazerte committed:
housekeeping

steffilazerte committed:
Pipe data through prepare_spatial

steffilazerte committed:
Update vignettes

steffilazerte committed:
Use correct syntax

steffilazerte committed:
Fix tests for init checks

steffilazerte committed:
loosen restrictions

steffilazerte committed:
Add missing snaps

steffilazerte committed:
Set seed with model_params() for GAM basis

steffilazerte committed:
Split year_basis out of snapshots for CI tests

steffilazerte committed:
Add CV support
- Also support for overwriting model output

steffilazerte committed:
Add missing documentation for CV

steffilazerte committed:
Clean up vignettes

steffilazerte committed:
Skip first diff models on win and mac for reasons :facepalm:

steffilazerte committed:
Clean up

steffilazerte committed:
Support passing all variables

steffilazerte committed:
Fine tune docs

AdamCSmithCWS committed:
insert st_make_valid into assign prov state

steffilazerte committed:
Add experimental note to Cross folds

steffilazerte committed:
Use wrapper snapshot function to make interactive testing easier

steffilazerte committed:
During stratify() add option to return omitted route-years if requested

steffilazerte committed:
Check for rnaturalearthhires too and use custom installation instructions

steffilazerte committed:
Explicitly add rnaturalearth hires

steffilazerte committed:
Housekeeping

steffilazerte committed:
Merge pull request #3 from AdamCSmithCWS/master

insert st_make_valid into assign prov state

steffilazerte committed:
See bbsBayes2!

steffilazerte committed:
Merge branch 'master' of github.com:steffilazerte/bbsBayes

steffilazerte committed:
Correct typo

steffilazerte committed:
again
